pkg/storage/bucket/client.go

```diff
@@ -117,26 +179,35 @@ func NewClient(ctx context.Context, cfg Config, name string, logger log.Logger,
 		return nil, err
 	}
 
-	client = opentracing.WrapWithTraces(bucketWithMetrics(client, name, reg))
+	if cfg.StoragePrefix != "" {
+		client = NewPrefixedBucketClient(client, cfg.StoragePrefix)
+	}
+
+	instrumentedClient := objstoretracing.WrapWithTraces(bucketWithMetrics(client, name, metrics))
 
 	// Wrap the client with any provided middleware
 	for _, wrap := range cfg.Middlewares {
-		client, err = wrap(client)
+		instrumentedClient, err = wrap(instrumentedClient)
 		if err != nil {
 			return nil, err
 		}
 	}
 
-	return client, nil
+	return instrumentedClient, nil
 }
 
-func bucketWithMetrics(bucketClient objstore.Bucket, name string, reg prometheus.Registerer) objstore.Bucket {
-	if reg == nil {
+func bucketWithMetrics(bucketClient objstore.Bucket, name string, metrics *Metrics) objstore.Bucket {
+	if metrics == nil {
 		return bucketClient
 	}
 
-	return objstore.WrapWithMetrics(
-		bucketClient,
-		prometheus.WrapRegistererWith(prometheus.Labels{"component": name}, reg),
-		"")
+	if metrics.BucketMetrics == nil {
+		reg := metrics.Registerer
+		reg = prometheus.WrapRegistererWithPrefix("loki_", reg)
+		reg = prometheus.WrapRegistererWith(prometheus.Labels{"component": name}, reg)
+		// Save metrics to be assigned to other buckets created with the same component name
+		metrics.BucketMetrics = objstore.BucketMetrics(reg, "")
```

Comment:
can we define the bucket metrics as a global in this file and register it with the default registerer during pkg init?
we can revise this in follow-up prs if needed. wdyt?